### ISSUE TITLE
Install the appropriate ChromeDriver version based on the currently-installed version of Chrome

### ISSUE THREAD

**Issue by Blackbaud-PaulCrowder** [angular/webdriver-manager#328]
A project I work on recently got bitten by a version mismatch between ChromeDriver and the locally-installed version of Chrome.  On my project, we have an automated build system that runs `webdriver-manager update` before running Selenium-based browser tests in Chrome so that the appropriate ChromeDriver bits are are installed.  However, we don't have full control over the version of Chrome on the machine where the tests are running, whether that's a hosted build agent in Microsoft VSTS or on an individual developer's machine.  As a result, the latest release of ChromeDriver was installed on a machine that had Chrome 68 installed, but the latest ChromeDriver is only compatible with Chrome 69+ and therefore caused the build to fail.  It would be great if `webdriver-manager update` had an option to install the ChromeDriver version that is compatible with the locally-installed version of Chrome so there's never a mismatch.

To get around the issue, I wrote a Node module that retrieves the local version of Chrome as well as the release notes for ChromeDriver, then parses the release notes and finds the version of ChromeDriver that's compatible with the installed version of Chrome.  I then pass this value to the `webdriver-manager update` command so that the correct ChromeDriver version is installed.  You can find the repo for this Node module [here](https://github.com/Blackbaud-PaulCrowder/chromedriver-version-matcher).  This solution is far from ideal as it depends on the release notes being formatted a specific way and hosted at specific URLs, so I'd be hesitant to say this exact solution is what should be implemented.  However, I was hoping that it might be inspiration to get some folks thinking about solving this issue in a more robust way.

**Comment by Blackbaud-PaulCrowder** [comment]
UPDATE: We ended up releasing an alpha version of the above-mentioned Node module, only this version uses a static list of compatible versions rather than parsing the release notes on every call.  We believe this iteration will be stable enough for broad adoption.  https://github.com/blackbaud/chromedriver-version-matcher

**Comment by cnishina** [comment]
We could use the json file however it requires maintenance. Do you want this to be a feature of webdriver-manager?

**Comment by cnishina** [comment]
Okay more thought into this then a question.

# Before Chrome browser 70

The ChromeDriver versions previous to version 70 required some mapping of ChromeDriver versions to Chrome browser versions. So for 69, you would need something like 2.43.

# After Chrome browser 70

It appears that there is also a ChromeDriver 70.0.3538.16. This, I am guessing works with Chrome version 70. Moving forward we could query the Chrome browser and figure out the appropriate version of ChromeDriver to download. Also because the way this project is downloading / parsing version 70, we are getting version 3538.16... and that is definitely wrong. I have a fix but I need to wait a few weeks before I'm back on this. 😄

**Comment by Blackbaud-PaulCrowder** [comment]
@cnishina Could you elaborate on "because the way this project is downloading / parsing version 70, we are getting version 3538.16"?  I just installed Chrome 70 on macOS, and when I run a test I get back ChromeDriver 2.43 as the matching version which is correct.  It's possible that if you're on a different platform that Chrome is reporting its version in a format that I'm not expecting, and that I might need to tweak the parsing logic to be more robust.  What OS are you on?

**Comment by cnishina** [comment]
Aaah. I found this issue when I was trying out my webdriver-manager-replacement. While I was making my side project, it was downloading 3538.16 because of regex. I fixed it and I thought this project might have parsed it the same way.

**Comment by Blackbaud-PaulCrowder** [comment]
OK.  So does your project attempt to solve the same problem as mine?

**Comment by cnishina** [comment]
No. I think this is a good thing for older Chrome versions. Moving forward, if the chromedriver team outputs a version that matches the browser major version, this might no longer be an issue.

**Comment by zwingliernst** [comment]
Not sure, if it fits here, but since I updated webdriver-manager to its latest version (due to this [issue 370](https://github.com/angular/webdriver-manager/issues/370) ), it downloads chromeDriver for Chrome Version 75. 
However, in my area of the world the latest Chrome Browser is Version 74.xyz
Can you fix this and align to use chromeDriver versions only from release Chrome Versions?

**Comment by cnishina** [comment]
The current version is 75. I just updated this to webdriver-manager@12.1.5.

**Comment by maplion** [comment]
@Blackbaud-PaulCrowder Thank you for taking this on; I'm running into this exact issue.  I think you're on the right track, but this definitely needs to be made more robust -- much too fragile, but thankfully, significantly less so with the tool.

**Comment by granttituss** [comment]
Thank you for your update. But maybe I do not understand something. I know that it is enough to visit the About Page to see the version of Chrome as it is described here [https://rocketfiles.com/](https://rocketfiles.com/articles/what-version-of-chrome-do-i-have). And you can check the version of driver and if it is compatible with yoru Chrome version.

**Comment by digeomel** [comment]
@Blackbaud-PaulCrowder your script works great still in 2020, thank you! 👍 😄

**Comment by Blackbaud-PaulCrowder** [comment]
Just FYI, my chromedriver-version-matcher has moved to a new repo and is now being published to NPM. It receives regular updates as new versions of Chrome are released.

[GitHub](https://github.com/blackbaud/chromedriver-version-matcher)
[NPM](https://www.npmjs.com/package/chromedriver-version-matcher)
